cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/GenesisKeyDelegationCertificate.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+
+module Cardano.CLI.EraBased.Run.Governance.GenesisKeyDelegationCertificate
+  ( runGovernanceGenesisKeyDelegationCertificate
+  ) where
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+import           Cardano.CLI.Types.Errors.GovernanceCmdError
+import           Cardano.CLI.Types.Key
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra
+
+runGovernanceGenesisKeyDelegationCertificate
+  :: ShelleyBasedEra era
+  -> VerificationKeyOrHashOrFile GenesisKey
+  -> VerificationKeyOrHashOrFile GenesisDelegateKey
+  -> VerificationKeyOrHashOrFile VrfKey
+  -> File () Out
+  -> ExceptT GovernanceCmdError IO ()
+runGovernanceGenesisKeyDelegationCertificate sbe
+                                             genVkOrHashOrFp
+                                             genDelVkOrHashOrFp
+                                             vrfVkOrHashOrFp
+                                             oFp = do
+  genesisVkHash <- firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrTextEnvFile AsGenesisKey genVkOrHashOrFp
+  genesisDelVkHash <-firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrTextEnvFile AsGenesisDelegateKey genDelVkOrHashOrFp
+  vrfVkHash <- firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrFile AsVrfKey vrfVkOrHashOrFp
+
+  req <- hoistEither $ createGenesisDelegationRequirements sbe genesisVkHash genesisDelVkHash vrfVkHash
+  let genKeyDelegCert = makeGenesisKeyDelegationCertificate req
+
+  firstExceptT GovernanceCmdTextEnvWriteError
+    . newExceptT
+    $ writeLazyByteStringFile oFp
+    $ shelleyBasedEraConstraints sbe
+    $ textEnvelopeToJSON (Just genKeyDelegCertDesc) genKeyDelegCert
+  where
+    genKeyDelegCertDesc :: TextEnvelopeDescr
+    genKeyDelegCertDesc = "Genesis Key Delegation Certificate"
+
+createGenesisDelegationRequirements
+  :: ShelleyBasedEra era
+  -> Hash GenesisKey
+  -> Hash GenesisDelegateKey
+  -> Hash VrfKey
+  -> Either GovernanceCmdError (GenesisKeyDelegationRequirements era)
+createGenesisDelegationRequirements sbe hGen hGenDeleg hVrf =
+  caseShelleyToBabbageOrConwayEraOnwards
+      (\eon -> return $ GenesisKeyDelegationRequirements eon hGen hGenDeleg hVrf)
+      (const $ Left GovernanceCmdGenesisDelegationNotSupportedInConway)
```

Comment:
So we don't want to fail here. What we want to happen is the parser to default to `Nothing` so the command isn't rendered in the first place.

You should change the first parameter to `ShelleyToBabbageEra era` here and call `caseShelleyToBabbageOrConwayEraOnwards` in the relevant parser.

Reply:
Have a look at: https://github.com/input-output-hk/cardano-cli/pull/369/commits/e49f027bc684c3ca9e8431683f8045a3d97a2305#diff-77a82b185df9945598785c342448c8a7d96017e485113dc9ccc4513b53d2647dR103

Reply:
@newhoggy> actually I have already constrained the non-legacy parser to not show the command for `conway`, via [forEraMaybeEon here](https://github.com/input-output-hk/cardano-cli/blob/f06df904d290f45ccaa5b4119ec58ca30beaf354/cardano-cli/src/Cardano/CLI/EraBased/Options/Governance.hs#L100) and the [ShelleyToAlonzo witness here](https://github.com/input-output-hk/cardano-cli/blob/f06df904d290f45ccaa5b4119ec58ca30beaf354/cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance.hs#L36).

But then I didn't change the `Run` part, because it is **shared with the legacy caller**, and I assumed I shouldn't change the legacy parser (which allows `conway` right now).

If you tell me I can constraint the legacy parser to rule out `conway`, then I can rule it out all the way, and so indeed remove the failure case you've spotted.

Let me know what I can do.

Reply:
> If you tell me I can constraint the legacy parser to rule out conway, then I can rule it out all the way, and so indeed remove the failure case you've spotted.

Let's try this and see how it turns out.

Reply:
Done :+1: I was able to make the legacy parser use the constraint `ShelleyToBabbage` instead of `ShelleyBasedEra` and rolling on from there, I was able to remove the offending failure case in `Run`.